unit/unit.go

```diff
@@ -0,0 +1,107 @@
+package unit
+
+import (
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"sourcegraph.com/sourcegraph/srclib/buildstore"
+)
+
+func init() {
+	buildstore.RegisterDataType("unit", SourceUnit{})
+}
+
+// UnitRepoUnresolved is a sentinel value that indicates this unit's
+// repository is unresolved
+const UnitRepoUnresolved = "?"
+
+func (u Key) IsResolved() bool {
+	return u.Repo == UnitRepoUnresolved
+}
+
+// ContainsAny returns true if u contains any files in filesnames. Currently
+// doesn't process globs.
+func (u SourceUnit) ContainsAny(filenames []string) bool {
+	if len(filenames) == 0 {
+		return false
+	}
+	files := make(map[string]bool)
+	for _, f := range filenames {
+		files[f] = true
+	}
+	for _, uf := range u.Files {
+		if files[uf] {
+			return true
+		}
+	}
+	return false
+}
+
+// idSeparator joins a source unit's name and type in its ID string.
+const idSeparator = "@"
+
+// ID returns an opaque identifier for this source unit that is guaranteed to be
+// unique among all other source units in the same repository.
+func (u SourceUnit) ID() ID {
+	return ID(fmt.Sprintf("%s%s%s", url.QueryEscape(u.Name), idSeparator, u.Type))
+}
+
+func (u *SourceUnit) ID2() ID2 {
```

Comment:
Agree, and if `ID2` is going to stay, it should be documented since it's an exported identifier (but primarily to help explain why it exists).